### ISSUE TITLE
False-positive incompatible type error without casting to defined Literal

### ISSUE THREAD

**Issue by tgelter** [python/mypy#9057]
* **Are you reporting a bug, or opening a feature request?**
Possible bug
* **Please insert below the code you are checking with mypy,**
```
#!/usr/bin/env python3

from typing import cast, Literal

valid_types = Literal["foo", "bar"]

def is_valid_type(my_type: valid_types) -> bool:
    if my_type == "foo":
        return True
    elif my_type == "bar":
        return True
    return False

def main() -> None:
    for t in ["foo", "bar"]:
        # If the following cast is removed, mypy reports an incompatible type:
        # Literal.py:19: error: Argument 1 to "is_valid_type" has incompatible type "str"; expected "Union[Literal['foo'], Literal['bar']]"
        t = cast(valid_types, t)
        if is_valid_type(t):
            print(f"{t} is valid")
        else:
            print(f"{t} is invalid")

if __name__ == "__main__":
    main()
```

* **What is the actual behavior/output?**
Error: `Literal.py:19: error: Argument 1 to "is_valid_type" has incompatible type "str"; expected "Union[Literal['foo'], Literal['bar']]"`
* **What is the behavior/output you expect?**
No error, since during each iteration of the loop, a matching string is found in the defined Literal 'valid_types', I believe the correctness of the invocation should be inferred, rather than reporting an error.
* **What are the versions of mypy and Python you are using?**
Python 3.8.3 / mypy 0.782
*  **Do you see the same issue after installing mypy from Git master?**
Unsure, `python3 -m pip install -U git+git://github.com/python/mypy.git` failed to clone when I tried.
* **What are the mypy flags you are using? (For example --strict-optional)**
None
* **If mypy crashed with a traceback, please paste the full traceback below.**
N/A

**Comment by JukkaL** [comment]
Mypy doesn't infer literal types unless context makes it apparent that a literal type should be required, since this would be too big of a compatibility break (and often not desirable). For example, consider code like this:

```py
x = ''
if cond():
    x += bar()
```

If `x` gets a literal type (empty string), the += operation would generate an error.

Literal types are pretty limited, unfortunately, and primarily useful in some fairly simple use cases.
